source/opt/cfg_cleanup_pass.cpp

```diff
@@ -0,0 +1,170 @@
+// Copyright (c) 2017 Google Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// This file implements a pass to cleanup the CFG to remove superfluous
+// constructs (e.g., unreachable basic blocks, empty control flow structures,
+// etc)
+
+#include <queue>
+#include <unordered_set>
+
+#include "cfg_cleanup_pass.h"
+
+#include "function.h"
+#include "module.h"
+
+namespace spvtools {
+namespace opt {
+
+void CFGCleanupPass::RemoveFromReachedPhiOperands(const ir::BasicBlock& block,
+                                                  ir::Instruction* inst) {
+  uint32_t inst_id = inst->result_id();
+  if (inst_id == 0) {
+    return;
+  }
+
+  analysis::UseList* uses = def_use_mgr_->GetUses(inst_id);
+  if (uses == nullptr) {
+    return;
+  }
+
+  for (auto u : *uses) {
+    if (u.inst->opcode() != SpvOpPhi) {
+      continue;
+    }
+
+    ir::Instruction* phi_inst = u.inst;
+    std::vector<ir::Operand> keep_operands;
+    for (uint32_t i = 0; i < phi_inst->NumOperands(); i++) {
+      const ir::Operand& var_op = phi_inst->GetOperand(i);
+      if (i >= 2 && i < phi_inst->NumOperands() - 1) {
+        // PHI arguments start at index 2. Each argument consists of two
+        // operands: the variable id and the originating block id.
+        const ir::Operand& block_op = phi_inst->GetOperand(i + 1);
+        assert(var_op.words.size() == 1 && block_op.words.size() == 1 &&
+               "Phi operands should have exactly one word in them.");
+        uint32_t var_id = var_op.words.front();
```

Comment:
It's a little more idiomatic t write:

```
 uint32_t var_id = GetSingleWordOperand(i);
 uint32_t block_id = GetSingleWordOperand(i+1);
```
rather than accessing the guts of an Operand instance.

Separately, both var_id and block_id should be const.